### PR TITLE
Viscous force and memory allocation

### PR DESCRIPTION
**Pull request by marinlauber**
This implements a simple viscous force routine. Following the structure of the routine for the pressure force, which needs to include the time of the simulation to make sure we are measuring at the correct time, see #110; 

Also, it is allocating like crazy, same as for the pressure force routine. 

This is a work in progress.

### CODE REVIEW

**Comment by marinlauber** [comment]
This fixes #110 by providing a higher-level interface function for the force routines. It also adds a moment computation routine. I will add some proper tests soon.

**Comment by marinlauber** [comment]
This is ready for review @weymouth @b-fg. I have added some examples I had lying around that show some of the capabilities added recently, but they didn't have an example. 

I also removed the type argument in `loc` and related functions.

File: src/util.jl [hunk 1]
```diff
@@ -121,8 +121,8 @@ using StaticArrays
 Location in space of the cell at CartesianIndex `I` at face `i`.
 Using `i=0` returns the cell center s.t. `loc = I`.
 """
-@inline loc(i,I::CartesianIndex{N},T=Float64) where N = SVector{N,T}(I.I .- 1.5 .- 0.5 .* δ(i,I).I)
-@inline loc(Ii::CartesianIndex,T=Float64) = loc(last(Ii),Base.front(Ii),T)
+@inline loc(i,I::CartesianIndex{N}) where N = SVector{N}(I.I .- 3//2 .- 1//2 .* δ(i,I).I)
```
Comment (b-fg) [inline src/util.jl:124]: Have you checked this works on GPU? It was not working for me on the AD branch.

File: src/util.jl [hunk 2]
```diff
@@ -33,6 +33,7 @@ function inside_u(dims::NTuple{N},j) where {N}
     CartesianIndices(ntuple( i-> i==j ? (3:dims[i]-1) : (2:dims[i]), N))
 end
 @inline inside_u(dims::NTuple{N}) where N = CartesianIndices((map(i->(2:i-1),dims)...,1:N))
+@inline inside_u(u::AbstractArray{T,N}) where {T,N} = CartesianIndices(map(i->(2:i-1),size(u)[1:end-1]))
```
Comment (b-fg) [inline src/util.jl:36]:
```suggestion
@inline inside_u(u::AbstractArray) = CartesianIndices(map(i->(2:i-1),size(u)[1:end-1]))
```

File: src/Metrics.jl [hunk 3]
```diff
@@ -80,11 +80,32 @@ end
...
+# viscous stress tensor
+∇²u(I::CartesianIndex{2},u) = @SMatrix [∂(i,j,I,u)+∂(j,i,I,u) for i ∈ 1:2, j ∈ 1:2]
+∇²u(I::CartesianIndex{3},u) = @SMatrix [∂(i,j,I,u)+∂(j,i,I,u) for i ∈ 1:3, j ∈ 1:3]
+"""
+   ∮τnds(u,body::AbstractBody,t=0)
+
+Compute the viscous force on a immersed body. 
+"""
+∮τnds(flow::Flow,body::AbstractBody) = ∮τnds(flow.u,body,time(flow))
+∮τnds(u::AbstractArray{T,N},body::AbstractBody,t=0) where {T,N} = sum(inside_u(u)) do I
```
Comment (b-fg) [inline src/Metrics.jl:97]:
```suggestion
∮τnds(u::AbstractArray,body::AbstractBody,t=0) = sum(inside_u(u)) do I
```

File: src/Metrics.jl [hunk 4]
```diff
@@ -80,11 +80,32 @@ end
 
 Surface normal integral of field `p` over the `body`.
 """
-function ∮nds(p::AbstractArray{T,N},df::AbstractArray{T},body::AbstractBody,t=0) where {T,N}
-    @loop df[I,:] .= p[I]*nds(body,loc(0,I,T),t) over I ∈ inside(p)
-    [sum(@inbounds(df[inside(p),i])) for i ∈ 1:N] |> Array
+∮nds(flow::Flow,body::AbstractBody) = ∮nds(flow.p,body,time(flow))
+∮nds(p::AbstractArray{T,N},body::AbstractBody,t=0) where {T,N} = sum(inside(p)) do I
```
Comment (b-fg) [inline src/Metrics.jl:84]:
```suggestion
∮nds(p::AbstractArray,body::AbstractBody,t=0) = sum(inside(p)) do I
```

**Comment by b-fg** [comment]
There are currently no tests for `∮τnds` and I could not get this function to work on a 2D flow. Have you checked that?

File: src/Metrics.jl (hunk 3 above)
Comment (b-fg) [reply src/Metrics.jl:97]: It's not `inside_u(u)` but `inside_u(size(u))` right?

Comment (marinlauber) [reply src/Metrics.jl:97]: No, that's why I had to write another function 'inside_u(u)` that points to the inside of the array, without the components indices, which would equate to `inside_u(size(u))[1:end-1]`. I just think it's cleaner that way.

File: src/util.jl (hunk 1 above)
Comment (marinlauber) [reply src/util.jl:124]: No, I have not

Comment (b-fg) [reply src/util.jl:124]: Ok, I probably won't. We are currently using `1.5f0`... because otherwise it doesn't work in GPUs.

Comment (marinlauber) [reply src/util.jl:124]: I will revert to the type conversion of the SVector for now. We can sort this out later.

**Comment by b-fg** [comment]
All tests passing locally on CPU and GPU.

@marinlauber the modifications I did to your previous changes (up to https://github.com/weymouth/WaterLily.jl/pull/113/commits/bb9959909cd466b7646387f44ca3eea388e388dd) make all the metrics work for GPU now, since that implementation was only working for CPU. The allocations are very good as well:
```
# Previous (bb99599)
# Pressure force: 0.040567 seconds (4 CPU allocations: 176 bytes)
# Viscous force: 0.343564 seconds (4 CPU allocations: 176 bytes)

# Current changes (d01f68c)
# Pressure force: 0.035714 seconds (39 CPU allocations: 1.562 KiB)
# Viscous force: 0.310422 seconds (38 CPU allocations: 1.516 KiB)
```
The only remaining thing to discuss is the examples added.  Only the [TwoD_cylinderVIV.jl](https://github.com/weymouth/WaterLily.jl/blob/d01f68c37dd94cfba55614ee37abb08b1941f07d/examples/TwoD_cylinderVIV.jl) had to be changed to comply with this PR, and the other examples do not actually even use any force/moment metric. I suggest we remove those changes from this PR and create a new one just for examples. Then this PR should be ready to merge.

**Comment by b-fg** [comment]
@weymouth can you please give it a look before we merge to master?

File: src/AutoBody.jl [hunk 5]
```diff
@@ -112,6 +112,11 @@ function measure(a::Bodies,x,t)
     sdf, map, _ = sdf_map_d(a.bodies,a.ops,x,t)
     measure(sdf,map,x,t)
 end
+# measures the distance, normal, and velocity accuratly for sdf≤1
+function measure_fast(body::AutoBody,x,t)
+    abs(body.sdf(x,t))>1 && return zero(eltype(x)),zero(x),zero(x)
```
Comment (weymouth) [inline src/AutoBody.jl:117]: Why doesn't this return `d, zero(x), zero(x)`?

**Comment by marinlauber** [comment]
Are we ok with discontinuing the `∮nds` function?

**Review by weymouth — COMMENTED** [review]
Looks good, except for that one default, which I can fix.

**Comment by b-fg** [comment]
> Are we ok with discontinuing the `∮nds` function?

Yeap, we will just keep the high-level functions since it was difficult to do multiple dispatch of `∮nds` applied to 2D/3D scalar/vector fields.

**Comment by b-fg** [comment]
> Looks good, except for that one default, which I can fix.

Which one?

File: src/AutoBody.jl (hunk 5 above)
Comment (marinlauber) [reply src/AutoBody.jl:117]: It's an error, should return d.

**Comment by b-fg** [comment]
The `promote_type` works but makes it a tiny bit slower even if there are less allocations (see tests above)
```
# Tests using SIMD execution (single-thread mode) and Array backend:

# Previous changes (d01f68c)
# Pressure force: 0.035714 seconds (39 CPU allocations: 1.562 KiB)
# Viscous force: 0.310422 seconds (38 CPU allocations: 1.516 KiB)

# Current changes (0f70067)
# Pressure force: 0.142078 seconds (19 CPU allocations: 800 bytes)
# Viscous force: 0.402344 seconds (19 CPU allocations: 768 bytes)
```
The price to pay for having these functions AD-compatible is fine though :)

**Pull request merged**